libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt

```diff
@@ -1,307 +1,690 @@
+            every { cancelRouteRefreshRequest(any()) } returns Unit
+        }
+        val routeRefreshController = createRouteRefreshController(
+            directionsSession = directionsSession
+        )
 
-        routeRefreshController.restart(validRoute) {}
-        coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
-        routeRefreshController.stop()
+        val refreshJob = async { routeRefreshController.refresh(listOf(createTestTwoLegRoute())) }
+        advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
+        refreshJob.cancel()
 
-        verify(exactly = 1) { directionsSession.cancelRouteRefreshRequest(requestId) }
+        verify(exactly = 1) { directionsSession.cancelRouteRefreshRequest(8) }
     }
 
     @Test
-    fun `do not send a request when uuid is empty`() {
-        every { routeOptions.enableRefresh() } returns true
-        every { validRoute.directionsResponse.uuid() } returns ""
+    fun `do not send a request when uuid is empty`() = coroutineRule.runBlockingTest {
+        val directionsSession = mockk<DirectionsSession>(relaxed = true)
+        val routeRefreshController = createRouteRefreshController(
+            directionsSession = directionsSession
+        )
+        val route = createTestTwoLegRoute(requestUuid = "")
 
-        routeRefreshController.restart(validRoute) {}
-        coroutineRule.testDispatcher.advanceTimeBy(routeRefreshOptions.intervalMillis * 2)
-        routeRefreshController.stop()
+        val refreshDeferred = launch { routeRefreshController.refresh(listOf(route)) }
+        coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
 
+        assertTrue(refreshDeferred.isActive)
         verify(exactly = 0) { directionsSession.requestRouteRefresh(any(), any(), any()) }
+        refreshDeferred.cancel()
+    }
+
+    @Test(expected = IllegalArgumentException::class)
+    fun `refresh route without legs`() = coroutineRule.runBlockingTest {
+        val initialRoute = createNavigationRoute(
+            createDirectionsRoute(
+                legs = null,
+                createRouteOptions(enableRefresh = true)
+            )
+        )
+        val routeRefreshController = createRouteRefreshController()
+
+        routeRefreshController.refresh(listOf(initialRoute))
     }
 
     @Test
-    fun `should cancel the previous request before starting a new one`() {
-        every { routeOptions.enableRefresh() } returns true
-
-        // Create 2 requests.
-        //   Let the first one create a requestId equal to 1.
-        //   Cancel during the second call, and then we expect the first request to be canceled.
-        val countDownLatch = CountDownLatch(2)
-        every { directionsSession.requestRouteRefresh(any(), any(), any()) } answers {
-            if (countDownLatch.count == 1L) {
-                routeRefreshController.restart(validRoute) {}
+    fun `should log route diffs when there is a successful response`() =
```

Comment:
Let's log incident diffs as well. Currently we're only logging annotations.

Reply:
This request feels not related to the cleanup feature 🤔 . Do you mind if I do this in the follow up PR after I merged this one?

Reply:
Sounds good :+1: